pallets/asset/src/identifier.rs

```diff
@@ -0,0 +1,173 @@
+use codec::{Decode, Encode};
+
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq)]
+pub enum Identifier {
+    CUSIP([u8; 9]),
+    CINS([u8; 9]),
+    ISIN([u8; 12]),
+    LEI([u8; 20]),
+}
+
+impl Identifier {
+    pub fn cusip(bytes: [u8; 9]) -> Option<Identifier> {
+        if cusip_checksum(&bytes[..8]) == bytes[8] - b'0' {
+            return Some(Identifier::CUSIP(bytes));
+        }
+        None
+    }
+
+    pub fn cins(bytes: [u8; 9]) -> Option<Identifier> {
+        if cusip_checksum(&bytes[..8]) == bytes[8] - b'0' {
+            return Some(Identifier::CINS(bytes));
+        }
+        None
+    }
+
+    pub fn isin(bytes: [u8; 12]) -> Option<Identifier> {
+        let s: String = bytes
+            .iter()
+            .map(|b| byte_value(*b))
+            .map(|b| b.to_string())
+            .collect();
+
+        let mut s1 = 0;
+        let mut s2 = 0;
+        for (i, c) in s.chars().rev().enumerate() {
+            let digit = c.to_digit(10)?;
+            if i % 2 == 0 {
+                s1 += digit;
+            } else {
+                s2 += 2 * digit;
+                if digit >= 5 {
+                    s2 -= 9;
+                }
+            }
+        }
+
+        if (s1 + s2) % 10 == 0 {
+            return Some(Identifier::ISIN(bytes));
+        }
+
+        None
+    }
+
+    pub fn lei(bytes: [u8; 20]) -> Option<Identifier> {
+        let mut s = bytes
+            .iter()
+            .take(18)
+            .map(|b| byte_value(*b))
+            .map(|b| b.to_string())
+            .collect::<String>();
+        s.push_str("00");
+        let check = 98 - (s.parse::<u128>().ok()? % 97);
+        if check == (((bytes[18] - b'0') * 10) + (bytes[19] - b'0')) as u128 {
+            return Some(Identifier::LEI(bytes));
+        }
+        None
+    }
+}
+
+fn cusip_checksum(bytes: &[u8]) -> u8 {
+    let mut v = 0;
+    let total = bytes.iter().enumerate().fold(0, |total, (i, c)| {
+        v = byte_value(*c);
+        if i % 2 != 0 {
+            v *= 2
+        }
+        total + (v / 10) + v % 10
+    });
+    (10 - (total % 10)) % 10
```

Comment:
```suggestion
    let total = bytes
        .iter()
        .copied()
        .map(byte_value)
        .enumerate()
        .map(|(i, v)| v << (i % 2))
        .map(|v| (v / 10) + v % 10)
        .sum()
    (10 - (total % 10)) % 10
```
A more purely functional variant of the algorithm; also `let mut v` is always overwritten on "loop"-entry, so having it outside has no effect as the previous loop-iteration value is not available.